### ISSUE TITLE
URLs over 1MB give "Out of memory" after connection has been made

### ISSUE THREAD

**Issue by jeroen** [curl/curl#12268]
### I did this

A user of the R bindings [has reported](https://github.com/jeroen/curl/issues/318) that URLs over 1MB in length always generate an oom error, after the request has completed.  He claims this started happening starting libcurl 7.71.0.

This does seem related to https://github.com/curl/curl/issues/6681 but it is maybe unexpected that this errors after the connection has been initiated? Perhaps if this is expected the error message can be improved.

Below a reprex in C:

```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>

int main(void)
{
  CURL *curl = curl_easy_init();
  char * url = calloc(1, 1300000);
  strcpy(url, "https://fakerapi.it/api/v1/custom?");
  for(int i = strlen(url); i < 1200000; i++){
    url[i] = 'x';
  }
  curl_easy_setopt(curl, CURLOPT_URL, url);
  curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
  curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
  CURLcode res = curl_easy_perform(curl);
  if(res != CURLE_OK)
    fprintf(stderr, "curl_easy_perform() failed: %s\n",
            curl_easy_strerror(res));
  return 0;
}
```

### curl/libcurl version

libcurl 7.71.0 and up.

### operating system

Anwy

**Comment by bagder** [comment]
It is a case of crappy error code + no good error message.

This code runs into the maximum size allowed for a HTTP request: 1024*1024 bytes (`DYN_HTTP_REQUEST`).

**Comment by anthonynorth** [comment]
Libcurl 7.70.0 and earlier had no URL length limit.

I agree the error message is misleading (thank you for fixing), but any error from curl here is unexpected. _Full URL syntax with no length limit_ is still documented as one of libcurl's features.

**Comment by bagder** [comment]
That phrase should probably updated/clarified.

Virtually no HTTP server will accept a request that large, which is one reason why libcurl no longer allows them. Also of course, no browser will ever use anything close to the size libcurl supports. libcurl does not actually limit the size of URLs but rather in some ways they are used, like HTTP. It can do longer requests for other protocols.

**Comment by bagder** [comment]
Of course @anthonynorth, if you have a genuine use case for a particular size we could discuss changing the maximum request size. The exact limit has been set a bit arbitrarily and I could certainly agree to a different value if there is a valid reason.

**Comment by anthonynorth** [comment]
My use case is genuine. The reprex I created for jeroen/curl#318 was a simple example reproducing the error and the threshold at which it occurs.

I'm using a web service which has no URL limitations (at least that we're aware of) and takes request parameters as path and query string. Largest requests that I can recall are ~66MB. My team have only hit this new limitation recently after updating to Ubuntu 2204, which ships with a newer curl (7.81.0).

Does the HTTP spec define a URL length limit?

> Also of course, no browser will ever use anything close to the size libcurl supports.
Browser URL limitations aren't relevant (in my opinion) to libcurl; usecases are different. Chrome has a 2MB limit according to this doco https://chromium.googlesource.com/chromium/src/+/master/docs/security/url_display_guidelines/url_display_guidelines.md#url-length

**Comment by jay** [comment]
Wow. Does it support POST?

**Comment by anthonynorth** [comment]
Unfortunately not.

**Comment by bagder** [comment]
> Largest requests that I can recall are ~66MB

Wow. That's... big. I'm reading this as a request to raise the limit. It has been 1MB since March 2021 and you are the first to run into this limit. Maybe we should just bump the limit to 100MB. But if we do that we also need to bump up the maximum string size we allow to *setopt() to 100MB - as that is 8MB right now.

I believe that is fine. Those limits are mostly there to prevent abuse and mistakes, but I think a 100MB limit can also work fine for that.

> Does the HTTP spec define a URL length limit?

This is not a URL limit you run into. This is a HTTP request size limit. Similar but not the same.

Then: no, it does not. But specs are specs and the real world is the real world. We need limits to reduce the risk of abuse and mistakes. There basically needs to be a limit, even if very big.

> Browser URL limitations aren't relevant (in my opinion) to libcurl

Back to the real world: browsers dictate how http and the web is used to a very large degree, so following and understanding what the browsers do and why is most certainly highly relevant. But they don't *decide* what curl should do or don't do.

> Chrome has a 2MB limit

Apparently, except that you can't enter such a long URL in its address bar which was the limit I think I thought of.

**Comment by bagder** [comment]
@anthonynorth I presume you would like to see #12279 merged then?

**Comment by anthonynorth** [comment]
> @anthonynorth I presume you would like to see https://github.com/curl/curl/pull/12279 merged then?

I've implemented a workaround, so I no longer need > 1MB requests. If no one else has hit this limitation, then it's probably not worth merging.

**Comment by bagder** [comment]
Thanks, yeah we will at least put that on hold if we don't get more people having problems. The error message should still be improved though.

**Comment by 3nippo** [comment]
Unfortunately, we also hit this limit (just a bit over 1MB), although I agree it is a result of bad design of our web app.

Error message "curl: (100) HTTP request too large" got me thinking it is http code 100 causing the problem, Google didn't help either. I figured out the cause only after some digging in the source code
